File: pydroid-ui/src/main/java/com/pyamsoft/pydroid/ui/loader/resource/RxResourceLoader.java
/*
 * Copyright 2016 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package com.pyamsoft.pydroid.ui.loader.resource;

import android.graphics.drawable.Drawable;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import com.pyamsoft.pydroid.helper.Checker;
import com.pyamsoft.pydroid.ui.loader.loaded.Loaded;
import com.pyamsoft.pydroid.ui.loader.loaded.RxLoaded;
import com.pyamsoft.pydroid.ui.loader.targets.Target;
import io.reactivex.Scheduler;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class RxResourceLoader extends ResourceLoader {

  @NonNull private final Scheduler obsScheduler;
  @NonNull private final Scheduler subScheduler;

  public RxResourceLoader(@DrawableRes int resource) {
    super(resource);
    obsScheduler = AndroidSchedulers.mainThread();
    subScheduler = Schedulers.io();
  }

  @NonNull @Override
  protected Loaded load(@NonNull Target<Drawable> target, @DrawableRes int resource) {
    final Target<Drawable> finalTarget = Checker.checkNonNull(target);
    return new RxLoaded(Single.fromCallable(() -> loadResource(appContext))
        .subscribeOn(subScheduler)
        .observeOn(obsScheduler)
        .doOnSubscribe(disposable -> {
          if (startAction != null) {
            startAction.call(finalTarget);
          }
        })
        .doOnError(throwable -> {
          Timber.e(throwable, "Error loading AsyncDrawable");
          if (errorAction != null) {
            errorAction.call(finalTarget);
          }
        })
        .doAfterSuccess(drawable -> {
          if (completeAction != null) {
            completeAction.call(finalTarget);
          }
        })
        .subscribe(finalTarget::loadImage,
            throwable -> Timber.e(throwable, "Error loading Drawable using RxResourceLoader")));
  }
}

File: pydroid-loader/src/main/java/com/pyamsoft/pydroid/loader/resource/RxResourceLoader.kt
/*
 * Copyright (C) 2018 Peter Kenji Yamanaka
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.pyamsoft.pydroid.loader.resource

import android.content.Context
import android.graphics.drawable.Drawable
import android.support.annotation.DrawableRes
import com.pyamsoft.pydroid.loader.cache.ImageCache
import com.pyamsoft.pydroid.loader.loaded.Loaded
import com.pyamsoft.pydroid.loader.loaded.RxLoaded
import com.pyamsoft.pydroid.loader.targets.Target
import io.reactivex.Scheduler
import io.reactivex.Single
import timber.log.Timber

internal class RxResourceLoader internal constructor(
  context: Context, @DrawableRes resource: Int, @DrawableRes errorResource: Int,
  resourceImageCache: ImageCache<Int, Drawable>,
  private val mainThreadScheduler: Scheduler,
  private val ioScheduler: Scheduler
) : ResourceLoader(context, resource, errorResource, resourceImageCache) {

  override fun load(target: Target<Drawable>, @DrawableRes resource: Int): Loaded {
    return RxLoaded(
        Single.fromCallable { loadResource() }
            .subscribeOn(ioScheduler)
            .observeOn(mainThreadScheduler)
            .doOnSubscribe { startAction?.invoke() }
            .doAfterSuccess { completeAction?.invoke(it) }
            .doOnError {
              Timber.e(it, "Error loading Drawable using RxResourceLoader")
              errorAction?.invoke(it)
            }.subscribe({ target.loadImage(it) }, { target.loadError(loadErrorResource()) })
    )
  }
}
